README.md
# tiopatinhas #

## Overview ##

*tiopatinhas* (tp) is a companion for AWS's Auto Scaling. It attaches itself to an
Availability Group and adds resources bought in Amazon's Spot market.

As we can lose those instances at any time (due to market conditions),
tiopatinhas only allows itself to provide about 50% of the total number of
running instances.

Other features include:

* efficient use of resources. tiopatinhas is aware of Amazon's one-hour
billing cycles and has protections against "flapping"
* crash recovery: in case there is a market crash, tiopatinhas can acquire
instances on the regular OnDemand market
* fail-safe design: tiopatinhas fails "up". If the process crashes, the worse
thing that can happen is extra servers being left in the cluster. Amazon's
Autoscaling rules are not modified at all and can take over at any time.

Our theory of operations is that even when the system is live, each server is
actually about 50% idle so that the system can responde to changes in access
patterns. How much you can commit to Spot Instances is correlated to how low you
keep your load during normal system operations. Needless to say, you shouldn't
use Spot Instances in systems which are not fault tolerant.

## Getting Started ##


### Before starting ###

1. Make sure the boto python package is installed in your system. If you
use debian or ubuntu you can install it by typing: 'sudo pip install boto'.
2. Make sure your AWS credentials are specified in a boto configuration file
(typically ~/.boto). Instruction on how to setup this file can be found [here](https://code.google.com/p/boto/wiki/BotoConfig).
3. To run the setup, you may need to install [setuptools](https://setuptools.readthedocs.io/en/latest/)

### Configuring tiopatinhas ###

* Copy the template conf file (tp.conf.template) to tp/tp.conf so that the
script can read it and make the changes according to your needs. Tio patinhas
currently supports the following properties:

#### Mandatory Properties

* *max_price:* A map that specifies the maximum bid prices for each type
  of EC2 instance. TP will use the prices specified in this map to bid for
  instances of that type in the spot market.
* *max_candidates:* The maximum number of instances TP will manage.
* *instance_name:* The prefix that will be used by TP to name managed instances.
* *region:* The AWS region where the AutoScaling group is located.
* *placement:* The AWS availability zone where TP instances will be launched.

#### Advanced properties

* *spot_type:* The instance type to bid for in the spot market (recommended: the same instance from the ASG)
* *emergency_type:* The instance type to buy in case of market crash (recommended: the same instance from the ASG)
* *bid_threshold:* Time to wait before doing another spot bid to AWS. Defaults to 300 seconds.
    * More information can be found [here](https://aws.amazon.com/ec2/spot/pricing/).
* *cool_down_threshold:* Time to wait before doing another scale action again. Defaults to 360 seconds.

#### Optional properties
* *tags:* A map containing custom metadata tags that must assigned to TP instances. *(optional)*
    * More information can be found [here](http://docs.aws.amazon.com/AWSEC2/latest/UserGuide/Using_Tags.html).
* *user_data_file:* An optional script or data that will be supplied to the instance on startup.
  If not provided, it will try to get from the Launch Configuration Group. *(optional)*
    * More information can be found [here](http://docs.aws.amazon.com/AWSEC2/latest/UserGuide/AESDG-chapter-instancedata.html).
* *subnet_id:* The VPC's subnet id that will be used by instances TP will manage. *(optional)*
    * More information can be found [here](http://docs.aws.amazon.com/AmazonVPC/latest/UserGuide/VPC_Subnets.html).
* *monitoring_enabled:* There are two types of Monitoring: basic and detailed.
  You can enable or disable the detailed monitoring by setting this field to True or False. *(optional)*
    * More information can be found [here](https://aws.amazon.com/cloudwatch/details/#amazon-ec2-monitoring).

### Coding with tiopatinhas  ###

To install the latest version directly from [GitHub](https://github.com/chaordic/tiopatinhas):

```bash
$ git clone https://github.com/chaordic/tiopatinhas.git
$ python tiopatinhas/setup.py install
```

You may need to use `sudo` depending on your environment setup.

Then:

```python
from tp import tp

t = tp.TPManager("auto scaling group name", debug=verbose)
t.run()
```

### Executing tiopatinhas ###

To install the latest version directly from [GitHub](https://github.com/chaordic/tiopatinhas):

```bash
$ git clone https://github.com/chaordic/tiopatinhas.git
$ python tiopatinhas/setup.py install
```

You may need to use `sudo` depending on your environment setup.

* Once the tp/tp.conf file is ready, execute tiopatinhas with the following command:
    * _python tp.py -g \<AutoScalingGroupName\>_ (this command must currently be executed from within the "tp" folder)
* You must optionally supply options "-v" for verbose mode or "-d" for daemon mode.

## License

tiopatinhas is available under the [MIT license](LICENSE).

CHANGELOG.md
## 1.0.3 (January 26, 2017)
    - Cool down no longer affects the tiopatinhas target anymore, target is always updated
    - Cool down is now added to instance's promotion and demotion

## 1.0.2 (December 02, 2016)
    - Added cool down and bid threshold times in the configuration file
    - Refresh state doesn't affect tiopatinhas' actions anymore

## 1.0.1 (November 30, 2016)
    - Added support to get user data from Launch Configuration Group if not provided

## 1.0.0 (October 11, 2016)
    - Added Monitoring enabled option
    - Added submodule support
    - Added VPC support
    - Added instance IAM role/profile support
    - Removed CPU checks from tiopatinhas and added weight factor (now you decide the ratio of machines that tiopatinhas will handle and the ASG handles everything else)
    - Fixes on Market crashed handlers and LB bugs
    - Instances marked-for-termination are now considered already down

## 0.1.1 (November 21, 2013)

Features:
    - Added spot and emergency type configuration (advanced feature that allows to use instance types different from the ASG)
    - Enable cloudwatch spot instances monitoring

Bugfixes:
    - Added fault tolerance to HTTP request errors (will now log error instead of crashing)

## 0.1.0 (August 27, 2013)

Features:

    - Configured logging per TPManager instance (instead of global logging)
    - Added ability to stop a TPManager instance
    - Enabled region-awareness (specified via config file or parameter)
    - Added support to tags (specified via config file)
    - Extracted user data to configuration file (or parameter)
    - Enabled verbose (debug) mode

Bugfixes:

    - Fixed bug that was preventing market crash to be detected when there were active/open bids.
    - Dettaching emergency instances from Load Balancer when they are demoted.


tp/cw.py
#!/usr/bin/env python

import datetime
import boto


class TendenceGuesser:
    def __init__(self):
        pass

    def guess(self):
        pass

class CPUTendenceGuesser:
    def __init__(self, ag_name, lower, upper, lower_threshold, upper_threshold):
        self.cloudwatch = boto.connect_cloudwatch()
        self.upper = upper - 3
        self.lower = lower + 3
        self.lower_threshold = lower_threshold
        self.upper_threshold = upper_threshold
        self.ag_name = ag_name

    def avg(self, l):
        return sum(l) / len(l)

    def guess(self, last_action=None):
        end = datetime.datetime.utcnow()
        start = end - datetime.timedelta(minutes=10)

        stats = self.cloudwatch.get_metric_statistics(60, start, end, 'CPUUtilization', 'AWS/EC2', 'Maximum', {'AutoScalingGroupName': self.ag_name})
        raw_stats = [x['Maximum'] for x in stats]

        half = len(raw_stats) / 2
        old = raw_stats[:half]
        new = raw_stats[half:]

        if len(filter(lambda x: x > self.upper, new)) > self.upper_threshold:
            return 1

        if len(filter(lambda x: x < self.lower, new)) > self.lower_threshold:
            return -1

        return 0


tp/tp.py
#!/usr/bin/env python

import boto
import time
import os
import logging
import sys
import simplejson as json
from collections import defaultdict
from boto import ec2
from boto.ec2 import autoscale
from boto.ec2 import elb
from boto.exception import EC2ResponseError
from itertools import chain
from datetime import timedelta
from datetime import datetime

logging.basicConfig(format='%(asctime)s %(name)s %(levelname)s %(message)s')
logger = logging.getLogger("main")
logger.setLevel(logging.DEBUG)


class AutoScaleInfoException(Exception):
    pass


class AutoScaleInfo:
    def __init__(self, autoscale_group_name, region):
        self.autoscale = boto.ec2.autoscale.connect_to_region(region)
        ags = self.autoscale.get_all_groups()
        self.name = autoscale_group_name

        try:
            self.ag = [x for x in ags if x.name == self.name][0]
        except:
            raise ValueError("Couldn't retrieve autoscale group info for %s" % autoscale_group_name)

        try:
            lcs = self.autoscale.get_all_launch_configurations(names=[self.ag.launch_config_name])
            self.lc = lcs[0]
        except:
            raise ValueError("Couldn't retrieve LaunchConfiguration for %s" % autoscale_group_name)

        self.instance_type = self.lc.instance_type
        self.image_id = self.lc.image_id
        self.security_groups = self.lc.security_groups
        self.user_data = self.lc.user_data

        self.load_balancers = self.ag.load_balancers
        self.desired_capacity = self.ag.desired_capacity

    def __repr__(self):
        return "<AutoScaleInfo Group:%s>" % self.name


class TPManager:
    def __init__(self, side_group, weight_factor=1.0, debug=False,
                 region=None, user_data=None, conf_file="tp.conf", az=None,
                 spot_type=None, grace_period_minutes=10):
        self.logger = logging.getLogger(side_group)
        if debug:
            self.logger.setLevel(logging.DEBUG)
        else:
            self.logger.setLevel(logging.INFO)

        self.conf = defaultdict(dict)
        try:
            with open(conf_file, 'r') as f:
                self.conf = json.loads(f.read())
        except IOError:
            self.logger.error("Configuration file " + conf_file + " not found.")
            sys.exit(2)

        self.grace_period_minutes = grace_period_minutes
        self.max_price = self.conf.get("max_price", {"c1.xlarge": "0.750"})
        self.spot_type = spot_type or self.conf.get("spot_type", "c1.xlarge")
        self.emergency_type = self.conf.get("emergency_type", "c1.xlarge")
        self.weight_factor = weight_factor
        self.tags = self.conf.get("tags", {})
        self.instance_profile_name = self.conf.get("instance_profile_name", None)
        self.region = region or self.conf.get("region", "us-east-1")  # parameter has precedence over config file
        self.subnet_id = self.conf.get("subnet_id", None)
        self.monitoring_enabled = self.conf.get("monitoring_enabled", False)
        self.cool_down_threshold = self.conf.get('cool_down_threshold', 360)
        self.bid_threshold = self.conf.get('bid_threshold', 300)

        if self.subnet_id is not None:
            self.placement = None
        elif az:
            self.placement = self.region + az
        else:
            self.placement = self.conf.get("placement", "us-east-1c")
        self.side_group = side_group
        self.tapping_group = AutoScaleInfo(self.side_group, self.region)

        self.started = False
        self.target = None
        self.last_bid = 0
        self.last_change = 0
        self.previous_as_count = None
        self.previous_managed = 0

        self.bids = []
        self.live = []
        self.emergency = []
        self.unhealthy_ids = set()

        self.ec2 = boto.ec2.connect_to_region(self.region)
        self.elb = boto.ec2.elb.connect_to_region(self.region)

        self.user_data = user_data
        user_data_file = self.conf.get("user_data_file", None)

        if not self.user_data and user_data_file:
            self.logger.info("Trying to get user data from file...")
            try:
                with open(user_data_file) as f:
                    self.user_data = f.read()
            except IOError:
                self.logger.warn("Could not read user data file: %s. Will launch instances without user data.",
                                 user_data_file)

        if not self.user_data:
            self.logger.info("Trying to get user data from launch configuration group...")
            self.user_data = self.tapping_group.user_data

        if not self.user_data:
            self.logger.warn("Could not read user from launch configuration group: %s."
                             " Will launch instances without user data.", self.tapping_group.lc.name)

        self.logger.info("User data: \n%s", self.user_data)

    def refresh(self):
        self.tapping_group = AutoScaleInfo(self.side_group, self.region)
        self.guess_target()
        if self.previous_as_count != self.managed_by_autoscale():
            self.logger.info(">> refresh(): autoscale instance count changed from %s to %s",
                             self.previous_as_count, self.managed_by_autoscale())
            self.previous_as_count = self.managed_by_autoscale()

    def is_ec2_state_running(self, instance_id):
        try:
            found_instance = self.ec2.get_all_instance_status(instance_ids=[instance_id])
            return len(found_instance) > 0 and found_instance[0].state_name == "running"
        except EC2ResponseError, inst:
            if inst.error_code == "InvalidInstanceID.NotFound":
                self.logger.warn("LB with invalid instance: %s", instance_id)
                return False
            else:
                raise inst

    def guess_target(self):
        if not self.started:
            self.target = min(self.managed_instances(), self.managed_by_autoscale())  # follow autoscale if stopped :)
            return

        if not self.target:
            self.target = self.managed_instances()
        previous = self.target

        # How many instances we should keep running
        candidate = round(self.weight_factor * self.tapping_group.desired_capacity)

        # Never less than one
        candidate = max(1, min(candidate, self.conf.get("max_candidates", 6)))

        if candidate != previous:
            self.logger.debug(">> guess_target(): changed target from %s to %s", previous, candidate)
            self.target = candidate

    def managed_by_autoscale(self):
        return int(self.tapping_group.desired_capacity)

    @property
    def lbs(self):
        lbnames = self.tapping_group.load_balancers
        lbs = self.elb.get_all_load_balancers(load_balancer_names=lbnames)
        return lbs

    def valid_bids(self):
        return [x for x in self.bids if x.state in ('active', 'open')]

    def managed_instances(self):
        return len(self.valid_bids()) + len(self.live) + len(self.emergency)

    def live_or_emergency(self):
        return len(self.live) + len(self.emergency)

    def ready_instances(self):
        return [x for x in self.bids if x.state == 'active']

    def buy(self, amount=1):
        tapping_group = self.tapping_group

        ami = self.ec2.get_image(tapping_group.image_id)
        for c in range(amount):
            r = ami.run(security_group_ids=tapping_group.security_groups,
                        instance_type=self.emergency_type,
                        instance_profile_name=self.instance_profile_name,
                        placement=self.placement,
                        subnet_id=self.subnet_id,
                        user_data=self.user_data,
                        monitoring_enabled=self.monitoring_enabled)

            self.logger.info(">> buy(): purchased 1 on-demand instance")
            time.sleep(3)

            # get instance since count is 1
            instance = r.instances[0]

            # get the status
            status = instance.update()

            # check for the status while isn't running
            while status != 'running':
                time.sleep(5)
                status = instance.update()

            # when it's finally running, update the tag
            self.logger.info("Instance status = " + status)
            instance.add_tag('tp:group', tapping_group.name)

    def bid(self, force=False):
        elapsed_time = time.time() - self.last_bid
        if not force and elapsed_time < self.bid_threshold:
            self.logger.info(">> bid(): last bid was too recent, skipping bid! Remaining time to next change %s",
                             self.bid_threshold - elapsed_time)
            time.sleep(10)
            return

        tapping_group = self.tapping_group

        request = self.ec2.request_spot_instances(
            price=self.max_price[self.spot_type],
            image_id=tapping_group.image_id,
            count=1,
            type="one-time",
            placement=self.placement,
            security_group_ids=tapping_group.security_groups,
            subnet_id=self.subnet_id,
            user_data=self.user_data,
            instance_type=self.spot_type,
            instance_profile_name=self.instance_profile_name,
            monitoring_enabled=self.monitoring_enabled)

        # get the first reservation since count is 1
        reservation = request[0]

        # get the status
        status = reservation.status

        # check for the status while it isn't running
        while status.code != 'fulfilled':
            self.logger.info("Reservation status = %s", status.code)
            time.sleep(5)
            status = self.ec2.get_all_spot_instance_requests(request_ids=[reservation.id])[0].status

        # when it's finally fulfilled, update the tag
        self.logger.info("Reservation status = %s", status.code)
        reservation.add_tag('tp:tag', self.side_group)

        self.logger.info(">> bid(): created 1 bid of %s for %s", self.spot_type, self.max_price[self.spot_type])
        self.last_bid = time.time()
        self.bids.append(request)

    def check_alive(self, instance_id):
        all_instances = self.ec2.get_all_instances(instance_ids=[instance_id])
        inst = all_instances.pop().instances[0]
        return inst.state == "running"

    def attach_instance(self, instance_id, infix):
        tags = self.tags.copy()
        tags['Name'] = "%s %s %s" % (self.conf.get("instance_name", "instance"), infix, self.side_group)
        self.ec2.create_tags([instance_id], tags)

        for lb in self.lbs:
            lb.register_instances(instance_id)

    def dettach_instance(self, instance_id):
        for lb in self.lbs:
            lb.deregister_instances(instance_id)

    def maybe_terminate(self, instance_id):
        # only if it's a spot or emergency machine, otherwise AS will take care of it
        live_spot_dict = dict([(x.instance_id, x) for x in self.live])
        live_emergency_ids = [x.id for x in self.emergency]

        if instance_id in live_spot_dict or instance_id in live_emergency_ids:
            # check grace period
            grace_period_delta = timedelta(minutes=self.grace_period_minutes)

            instance_info = self.ec2.get_all_instances(instance_ids=[instance_id])[0].instances[0]
            uptime = datetime.utcnow() - datetime.strptime(instance_info.launch_time, '%Y-%m-%dT%H:%M:%S.%fZ')
            if uptime > grace_period_delta:
                self.logger.info(">> maybe_terminate(): %s is unhealthy for longer than %s minutes - killing it!",
                                 instance_id, self.grace_period_minutes)
                self.dettach_instance(instance_id)
                self.ec2.terminate_instances([instance_id])

                if instance_id in live_spot_dict:
                    instance = live_spot_dict[instance_id]
                    instance.cancel()  # cancel the spot request
                    self.live.remove(instance)

    def maybe_promote(self, spot_request):
        elapsed_time = time.time() - self.last_change
        if elapsed_time <= self.cool_down_threshold:
            self.logger.info("Not promoting any instances, waiting for cool down!"
                             " Remaining time to next change %s", self.cool_down_threshold - elapsed_time)
        elif self.check_alive(spot_request.instance_id):
            self.logger.info(">> maybe_promote(): %s is alive, promoting", spot_request)

            self.attach_instance(spot_request.instance_id, "TP")
            self.bids.remove(spot_request)
            self.live.append(spot_request)
            self.last_change = time.time()
            self.logger.info(">> maybe_promote(): %s promoted, now live", spot_request)

    def maybe_replace(self):
        for instance in self.emergency:
            self.logger.debug("proximity(%s): %s", instance.id, str(self.proximity(instance)))
            if (2 < self.proximity(instance) < 10) and self.managed_instances() <= self.target:
                self.logger.info(">> maybe_replace(): attempting to replace %s", instance.id)
                self.bid(force=True)

            self.load_state()

    def proximity(self, instance_or_spot):
        if hasattr(instance_or_spot, 'instance_id'):
            instance_info = self.ec2.get_all_instances(instance_ids=[instance_or_spot.instance_id])[0]
            instance = instance_info.instances[0]
        else:
            instance = instance_or_spot

        minute = int(instance.launch_time.split(":")[1])
        minute_now = datetime.now().minute
        time_remaining = minute - minute_now
        if time_remaining < -1:
            time_remaining = (minute + 60) - minute_now
        return time_remaining

    def maybe_demote(self):
        # First remove open, unfulfilled bids
        # Then remove open, but not yet live
        # Finally, remove any

        # In case we are in an emergency state:
        # If we have a server in an emergency state and no bids are open, kill a
        # server
        if self.emergency:
            for instance in self.emergency:
                if (3 < self.proximity(instance) < 10) and not self.valid_bids():
                    self.logger.info(">> maybe_demote(): removing emergency instance %s", instance.id)
                    self.dettach_instance(instance.id)
                    self.ec2.terminate_instances([instance.id])
                    self.emergency.remove(instance)
                    return True
            return False

        if self.managed_instances() <= self.target:
            return False

        for bid in self.valid_bids():
            if bid.state == "open":
                self.logger.info(">> demote(): %s is open, removing", bid)
                bid.cancel()
                self.bids.remove(bid)
                return True

        elapsed_time = time.time() - self.last_change

        if elapsed_time <= self.cool_down_threshold:
            self.logger.info("Not removing any instances, waiting for cool down!"
                             " Remaining time to next change %s", self.cool_down_threshold - elapsed_time)
            return False

        self.live.sort(key=self.proximity)

        if self.proximity(self.live[0]) < 3 or not self.started:
            self.logger.info(">> demote(): %s is live, removing", self.live[0])
            self.dettach_instance(self.live[0].instance_id)
            self.last_change = time.time()
            time.sleep(5)
            self.live[0].cancel()
            self.ec2.terminate_instances([self.live[0].instance_id])
            time.sleep(1)
            self.live.remove(self.live[0])
            return True
        else:
            self.logger.info(">> demotion too far off, postponing (%s minutes)", self.proximity(self.live[0]))

        return False

    def load_state(self):
        running_in_lb = []
        self.unhealthy_ids = set()

        for lb in self.lbs:
            for instance_state in lb.get_instance_health():
                if instance_state.state != 'InService':
                    self.unhealthy_ids.add(instance_state.instance_id)
                # Some times some dead instances get stuck on LB and boto lib doesn't know how to treat it
                # This make sure that instance is alive and avoid bug on get_all_instances method
                elif self.is_ec2_state_running(instance_state.instance_id):
                    running_in_lb.append(instance_state.instance_id)
                else:
                    self.dettach_instance(instance_state.instance_id)

        spot_requests = self.ec2.get_all_spot_instance_requests()
        self.bids = []
        self.live = []

        for request in spot_requests:
            tp_tag = request.tags.get('tp:tag', None)
            if not tp_tag or tp_tag != self.side_group:
                continue

            if request.instance_id not in running_in_lb:
                self.bids.append(request)
            elif request.status.code != 'marked-for-termination':
                self.live.append(request)

        self.emergency = []

        # gets a list of instances and flat the items from all element.instances
        all_instances = [r.instances for r in self.ec2.get_all_instances()]
        instances = chain.from_iterable(all_instances)
        for instance in instances:
            if instance.tags.get('tp:group', None) == self.tapping_group.name and instance.state == "running":
                self.emergency.append(instance)
                if instance.id not in running_in_lb:
                    self.logger.info(">> load_state: Attaching new emergency instance %s to LB." % instance.id)
                    self.attach_instance(instance.id, "OD")

    def stop(self):
        """ Prepares this TPManager to stop by not launching new machines
            and gradually remove old machines.

            This manager loop will only stop when both the autoscaling group
            and the TP manager has zero instances running.
        """
        self.started = False

    def start(self):
        self.started = True

    def print_state(self):
        self.logger.debug("*** Current state:")
        self.logger.debug("Managed by Autoscale: " + str(self.managed_by_autoscale()))
        self.logger.debug("TP target: " + str(self.target))
        self.logger.debug("Instances managed by TP: " + str(self.managed_instances()))
        self.logger.debug("TP live instances: " + str(len(self.live)) + " [" +
                          ", ".join([x.instance_id for x in self.live]) + "]")
        self.logger.debug("Emergency: " + ", ".join([x.id for x in self.emergency]))
        self.logger.debug("LB Unhealthy: " + ", ".join(self.unhealthy_ids))

    def run(self):
        self.start()
        self.previous_managed = 0
        self.logger.info("Starting Tio Patinhas")
        while self.started or self.managed_instances() > 0:
            try:
                self.save_money()
            except Exception, e:
                logger.exception(e)
                time.sleep(10)

            flush_output()
            time.sleep(20)
        self.logger.debug("Stopped running.")

    def save_money(self):
        self.logger.debug("Refreshing state...")
        self.load_state()
        self.refresh()
        self.print_state()

        self.logger.debug("Checking if needs to launch emergency instances")
        if self.started and self.previous_managed > 0 and self.live_or_emergency() == 0:
            self.logger.warn(">> market crashed! launching %s %s instances", self.previous_managed, self.emergency_type)
            self.buy(self.previous_managed)
            self.load_state()

        self.logger.debug("Checking if there's any emergency instance to replace")
        if self.emergency:
            self.maybe_replace()

        self.logger.debug("Checking if it needs to buy spot instances")
        if self.managed_instances() < self.target:
            self.bid()
            self.load_state()

        self.logger.debug("Checking if there's any instance ready to be attached")
        for new in self.ready_instances():
            self.maybe_promote(new)

        self.logger.debug("Checking if there's any sick machine to terminate")
        for sick in self.unhealthy_ids:
            self.maybe_terminate(sick)

        self.maybe_demote()
        self.previous_managed = self.live_or_emergency()


def flush_output():
    sys.stdout.flush()
    sys.stderr.flush()


def daemonize():
    pid = os.fork()
    if pid > 0:
        sys.exit()

    os.chdir("/")
    os.setsid()
    pid = os.fork()
    if pid > 0:
        sys.exit()

    in_ = file("/dev/null", 'r')
    out = file("/dev/null", 'a+')
    error = file("/dev/null", 'a+')

    flush_output()

    os.dup2(in_.fileno(), sys.stdin.fileno())
    os.dup2(out.fileno(), sys.stdout.fileno())
    os.dup2(error.fileno(), sys.stderr.fileno())


if __name__ == '__main__':
    import getopt


    def usage():
        print """\
Usage: tp [OPTIONS]

tp is a long-standing process that attaches itself to an availability group on
AWS. It attempts to buy cheap instances on the Spot Market and add those to the
availability's group load balancer.

   -g, --group                     Availability group to attach to
   -d, --daemonize                 Detach from the terminal
   -v, --verbose                   Verbose mode
"""


    try:
        opts, args = getopt.getopt(sys.argv[1:], "g:dv", ["group=", "daemonize", "verbose"])
    except getopt.GetoptError, err:
        logger.error(str(err))
        usage()
        sys.exit(2)

    group = None
    do_daemonize = False
    verbose = False

    for o, a in opts:
        if o in ("-g", "--group"):
            group = a
        elif o in ("-v", "--verbose"):
            verbose = True
        elif o in ("-d", "--daemonize"):
            daemonize = True
        else:
            assert False, "Unhandled option"

    if do_daemonize:
        daemonize()

    if not group:
        logger.error("no autoscale group defined")
        usage()
        sys.exit(2)

    tp = TPManager(group, debug=verbose)
    tp.run()
